compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala

```diff
@@ -602,16 +613,84 @@ private class ExtractAPICollector(using Context) extends ThunkHolder {
     annots.toList
   }
 
+  /** Produce a hash for a tree that is as stable as possible:
+   *  it should stay the same across compiler runs, compiler instances,
+   *  JVMs, etc.
+   */
+  def treeHash(tree: Tree): Int =
+    import scala.util.hashing.MurmurHash3
+
+    def positionedHash(p: ast.Positioned, initHash: Int): Int =
+      p match
+        case p: WithLazyField[?] =>
+          p.forceIfLazy
+        case _ =>
+      // FIXME: If `p` is a tree we should probably take its type into account
+      // when hashing it, but producing a stable hash for a type is not trivial
+      // since the same type might have multiple representations, for method
+      // signatures this is already handled by `computeType` and the machinery
+      // in Zinc that generates hashes from that, if we can reliably produce
+      // stable hashes for types ourselves then we could bypass all that and
+      // send Zinc hashes directly.
+      val h = MurmurHash3.mix(initHash, p.productPrefix.hashCode)
+      iteratorHash(p.productIterator, h)
+    end positionedHash
+
+    def iteratorHash(it: Iterator[Any], initHash: Int): Int =
+      import core.Constants._
+      var h = initHash
+      while it.hasNext do
+        it.next() match
+          case p: Positioned =>
+            h = positionedHash(p, h)
+          case xs: List[?] =>
+            h = iteratorHash(xs.iterator, h)
+          case c: Constant =>
+            h = MurmurHash3.mix(h, c.tag)
+            c.tag match
+              case NullTag =>
+                // No value to hash, the tag is enough.
+              case ClazzTag =>
+                // Go through `apiType` to get a value with a stable hash, it'd
+                // be better to use Murmur here too instead of relying on
+                // `hashCode`, but that would essentially mean duplicating
+                // https://github.com/sbt/zinc/blob/develop/internal/zinc-apiinfo/src/main/scala/xsbt/api/HashAPI.scala
+                // and at that point we might as well do type hashing on our own
+                // representation.
+                // There's no test for this codepath because it seems that we
+                // never generate trees of the form `Literal(Constant(someClassType))`
+                // before Pickler currently.
```

Comment:
I found that synthesised ClassTags will use constants to represent classes, which appear in tasty

https://github.com/lampepfl/dotty/blob/82255021f822762316fb288eec5ba7a1568da1a3/compiler/src/dotty/tools/dotc/typer/Synthesizer.scala#L42

Reply:
interesting, let me see if I can get a test out of that...